## Course Title

### Command Line Tools for Genomic Data Science



## Course Instructor

### Liliana Florea, Ph.D.

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

### Course Description

This class will cover an introduction to command line tools for managing genomic data. It will start with basic Unix commands for managing files and directories, followed by a review of the main starndardized data formats and hands-on instruction on how to use popular tools such as Samtools, Bedtools, Bowtie, and application specific technologies such as Bcftools and Cufflinks. 

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Course Content

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------



### Unix Commands

**Week 1 Overview**

- Basic commands

- Pipes

In the first week of this course, you will learn many specific and useful Unix commands, which are essential for basic text-based file analysis. In our case, these files contain biological information such as genome sequences, stored in formats like FASTA or FASTQ. What I found most valuable about the first week of this course was the detailed explanation of each command by Professor Liliana Florea. After watching the tutorial portion of the first week, you will be required to answer ten questions in a quiz. Following the quiz, there is an exam, which I believe can be challenging to pass successfully. Overall, the first week of this course has been one of the most informative sections I've encountered so far, and I genuinely appreciate Professor Liliana Florea for all the effort she has put into this fantastic course.



---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------



### Sequences and Genomic Features

**Week 2 Overview**

Week 2 of this course is divided into two main parts, each focused on distinct aspects of biological data handling and analysis. The first part is dedicated to gaining familiarity with various biological file formats. This includes popular sequence storage files like FASTA and FASTQ, as well as more specialized formats used to store data generated by next-generation sequencing tools, such as GTF, SAM, and BED files. These formats are fundamental for understanding and working with biological data.
The second part of the week is centered around using essential bioinformatics tools like Samtools and Bedtools through the command line to analyze data stored in the formats discussed in the first part. This practical experience will equip you with the skills needed to manipulate and interpret biological data effectively.

The course begins with an introduction to basic molecular concepts. While this may be familiar territory for molecular biologists, it provides essential foundational knowledge for data scientists or informaticians without a biological background.
In the first main part of the week, you'll delve into the specifics of various biological file formats. These formats are crucial for comprehending the subsequent section and are widely used in the field of genomics.
The second main part of the week introduces two powerful software packages: Samtools and Bedtools. You'll learn how to use these tools to analyze and manipulate biological data efficiently. These packages offer a wide range of functionalities for handling genomic information.

Just like in the first week, this section includes quizzes and exams to assess your knowledge and understanding of the course material. These assessments are valuable for gauging your progress and reinforcing your learning.



-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------



### Alignment and sequence variation

**Week 3 Overview**

- Bowtie, BWA

- Samtools, Bcftools

- Application: variant calling

In this section, we delve into the critical process of alignment, a fundamental step in the analysis of DNA or RNA sequencing data generated by next-generation sequencing machines. This involves mapping the generated reads to a reference genome, or in some cases, performing de novo alignment without a reference genome. The objective is to understand sequence variations and identify genetic variants, an essential task in genomics research.

The initial part of this section provides an in-depth introduction to alignment. Next-generation sequencing data, often comprising millions of short reads, must be aligned accurately to a reference genome or assembled de novo. This alignment forms the foundation for downstream analyses, enabling us to extract meaningful biological insights.

We revisit the SAM (Sequence Alignment/Map) and BAM (Binary Alignment/Map) formats, which were introduced in the previous section of the course. This time, we delve deeper into these formats, understanding how they store the rich information generated by next-generation sequencing tools.

In this section, we introduce two powerful alignment toolsâ€”Bowtie and BWA. These tools are essential packages, widely used for aligning sequencing reads to reference genomes. You will learn how to install and utilize them within a Linux command line environment.

As a participant, you will have the opportunity to practice what you've learned. You'll align sample sequencing files either by downloading them (if you're using your own Linux system) or within a virtual environment, introduced at the beginning of the course. This hands-on experience is invaluable in building your practical skills.

Following the alignment, you will delve into the analysis of alignment results using Samtools and Bcftools. These packages are essential for manipulating and analyzing SAM/BAM files and for performing variant calling, a key aspect of genomics research.

To reinforce your understanding, each section concludes with a quiz.  The section culminates in an exam to assess your proficiency in alignment and variant calling.



----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------



### Transcriptomics: RNA-seq Analysis

**Week 3 Overview**

The final section of this course has been a captivating exploration of transcriptomics, a field dedicated to the study of RNA sequences, gene expression, and regulation. We have delved into powerful Linux-based tools and harnessed them to analyze RNA-seq data, gaining invaluable insights into gene expression dynamics.

**Tophat: Mapping RNA-seq Reads**

Our journey began with Tophat, a versatile tool designed to align RNA-seq reads generated by next-generation sequencing machines. We utilized Tophat to perform alignments, understanding the significance of accurately mapping these reads. Subsequently, we delved into the analysis of the alignment results, honing our skills in this essential step.

**Cufflinks: Assembly of RNA-seq Reads**

Next, we ventured into the world of Cufflinks, a tool used for the assembly of RNA-seq reads. By employing command-line techniques, we harnessed the power of Cufflinks to create meaningful assemblies from the generated data. Understanding how to analyze the results marked another milestone in our learning journey.

**Cuffdiff: Differential Gene Expression Analysis**

We then harnessed Cuffdiff, a valuable tool for comparing RNA-seq samples from different conditions. With Cuffdiff, we explored differential gene expression, enabling us to unravel the unique genetic responses in each sample. Analyzing the results provided crucial insights into the biology under investigation.

**IGV: Visualizing Differential Expression**

Our last tool in this section was IGV (Integrative Genomics Viewer). Unlike the command-line tools we used previously, IGV provided a graphical interface to visualize the results of differential gene expression. It allowed us to intuitively explore and comprehend the data, making it accessible and informative.







---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------



**Conclusion**

As we wrap up this four-week course, I want to express my gratitude to Professor Liliana Florea for imparting this fantastic and informative knowledge. It's been a fulfilling journey of hard work and dedication. I'm delighted to have completed the course with an excellent grade, and I look forward to applying the skills and knowledge gained in my future endeavors.



Loghman Samani

November 8, 2023













